enhancements/machine-api/control-plane-machine-set.md

```diff
@@ -0,0 +1,908 @@
+Due to this limitation, the CRD will be cluster scoped.
+
+The behaviour of such a controller is complex, and as such, various features of the controller and scenarios are
+outlined in the details below.
+
+#### Desired number of Machines
+
+At present, the only source of the installed control plane size within OpenShift clusters exists within the
+`cluster-config-v1` ConfigMap in the `kube-system` namespace.
+This ConfigMap has been deprecated for some time and as such, should not be used for scale in new projects.
+
+Due to this limitation, we will need to have a desired replicas field within the ControlPlaneMachineSet controller.
+As we are not planning to tackle horizontal scaling of the control plane initially, we will implement a validating
+webhook to deny changes to this value once set.
+
+For new clusters, the installer will create (timeline TBD) the `ControlPlaneMachineSet` resource and will set the value
+based on the install configuration.
+For existing clusters, we will need to validate during creation of the `ControlPlaneMachineSet` resource that the
+number of existing Control Plane Machines matches the replica count set within the `ControlPlaneMachineSet`.
+This will prevent end users from attempting to horizontally scale their control plane during creation of the
+`ControlPlaneMachineSet` resource.
+
+In the future, once we have identified risks and issues with horizontal scale, and mitigated those, we will remove the
+immutability restriction on the replica field to allow horizontal scaling of the control plane between 3 and 5 replicas
+as per current support policy for control plane sizing.
+
+#### Selecting Machines
+
+The ControlPlaneMachineSet operator will use the selector defined within the CRD
+to find Machines which it should consider to be within the set of control plane Machines it is responsible for managing.
+
+This set should be the complete control plane and there should be a 1:1 mapping of Machines in this set to the control
+plane nodes within the cluster.
+
+If there are any control plane nodes (identified by the node role) which do not have a Machine, the operator will mark
+itself degraded as this is an unexpected state.
+No further action will be taken until the unknown node has either been removed from the cluster or a Machine has been
+manually created for it.
+
+#### Providing high availability/fault tolerance within the control plane
+
+Typically within OpenShift, Machines are created within a MachineSet.
+MachineSets have a single `ProviderSpec` which defines the configuration for the Machines created by the MachineSet.
+The failure domain (sometimes called availability zone) is a part of this provider spec and as such, defines that all
+Machines within the Machineset share a failure domain.
+
+This is undesirable for Control Plane Machines as we wish to have them spread across multiple availability zones to
+reduce the likelihood of datacenter level faults degrading the control plane.
```

Comment:
How is this different from compute nodes, which customers also want to be spread across multiple availability zones, and where we meet this need by creating multiple machine sets with one for each failure domain?

Reply:
It's not any different, but this is particularly undesirable for the Control Plane IMO. 

With workloads, they are the customers responsibility and it is up to them to spread across multiple failure domains. The Control Plane is more in our hands so we should endeavour to make this more visible/obvious and implement the best practice here.

Reply:
Can these guard rails belong somewhere else? For example, the Installer could determine to provision multiple machinesets for control plane just like we currently do for compute. I strongly advise against duplicating code just to enforce "best practice" as those code paths are eventually going to diverge and provide inconsistent user experience.

Reply:
I've tried to highlight this within the alternatives section, but to summarise, there have been previous iterations at how we might achieve this using just regular MachineSets. With each previous iteration there were a huge number of risks and hacks that were going to be needed to make this safe, for example, what happens when a user deletes a machineset owning a control plane machine, the machine goes away and they lose a control plane member, is that something we want to allow customers to do? There are other examples but they're already documented. Having spoken with a few people, while I agree, the duplication is not ideal, there are also a lot of good reasons to have a separate, specific control plane management which I think outweigh the negative of duplication here